<a href="https://colab.research.google.com/github/Sagar-Gilda/Python/blob/main/5_Data_Cleaning_output.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center><u>Data Cleaning</u></center>

 Data cleaning is a part of the process on a data science project.
 
Data cleaning is the process of fixing or removing incorrect, corrupted, incorrectly formatted, duplicate, or incomplete data within a dataset.
<br>
When you clean your data, all outdated or incorrect information is gone – leaving you with the highest quality information for your analysis and model building.

<p style='text-align: right;'> 4 points</p>


In [ ]:
# import pandas and numpy with alias pd and np respectively
import pandas as pd
import numpy as np

Create a dataframe, d1 = pd.DataFrame( {‘Temperature’ : [1, np.nan, 3, 2, 3] ,’Humidity’ : [22, np.nan, 2 , np.nan, 20 ] })

In [ ]:
#create d1
d1 = pd.DataFrame( {'Temperature' : [1, np.nan, 3, 2, 3] ,'Humidity' : [22, np.nan, 2 , np.nan, 20 ] })

print the dataframe d1

In [ ]:
#print d1
d1

Temperature  Humidity
0          1.0      22.0
1          NaN       NaN
2          3.0       2.0
3          2.0       NaN
4          3.0      20.0

Find whether the given dataframe contain any missing values? 

In [ ]:
#check for null
d1.isnull()

Temperature  Humidity
0        False     False
1         True      True
2        False     False
3        False      True
4        False     False

How many missing values does each column have?

In [ ]:
#total null
d1.isnull().sum()

Temperature    1
Humidity       2
dtype: int64

## Dealing with missing values

Refrence:-https://www.youtube.com/watch?v=xkRz6R0FlQ4

<br>Now, we know we have missing values, the next thing that we need to work on, is how to deal with these missing values

 - ## Method 1: Delete the rows which contain missing values.
 This method include dropping all the rows that have missing value in any column. 

 <p style='text-align: right;'> 20 points</p>

Use a suitable method to drop all the rows having missing values and save the change in d2 variable

In [ ]:
d2=d1.dropna()  #drop nan

Print d2

In [ ]:
#print nan
d2

Temperature  Humidity
0          1.0      22.0
2          3.0       2.0
4          3.0      20.0

Hey Remember : droping rows with nan is one of the method to deal with missing values. But you have to decide if you need to go for this method by checking percentage of nan present in the dataframe.

If a column is having more than 60% of nan values then its better to remove such variables altogether if business permits 

- ## Method 2: Replacing missing values
Sometimes rather than dropping NA values, you'd rather replace them with a valid value. This value might be a single number like zero, or it might be some sort of imputation or interpolation

<p style='text-align: right;'> 30 points</p>


Impute the missing values with constant number of your choice

In [ ]:
# The below output has imputed missing  values with 100
d1.fillna(100)



Temperature  Humidity
0          1.0      22.0
1        100.0     100.0
2          3.0       2.0
3          2.0     100.0
4          3.0      20.0

Do you think its a  good way to treat Nan values? What if such constant values are not suitable for our further analysis?
Try to give your thoughts on this.

In [ ]:
#No Nan wont be an ideal way to impute values for continuous data attributes as it might not be suitable for data distribution 

Impute the missing values with mean

In [ ]:
#  imputing mean
d1.fillna(np.mean(d1))

Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  14.666667
4         3.00  20.000000

Impute the missing values with median

In [ ]:
#median imputing
d1.fillna(np.median(d1))

Temperature  Humidity
0          1.0      22.0
1          NaN       NaN
2          3.0       2.0
3          2.0       NaN
4          3.0      20.0

Replacing with the mean, mode or median approximations are a statistical approach of handling the missing values.

Another Fun fact:
    
    This is an approximation which can add variance to the data set. But the loss of the data can be negated by this method which yields better results compared to removal of rows and columns.

Impute Nan with forward fill

In [ ]:
#forward fill
d1.fillna(method='ffill')

Temperature  Humidity
0          1.0      22.0
1          1.0      22.0
2          3.0       2.0
3          2.0       2.0
4          3.0      20.0

Impute Nan with backward fill

In [ ]:
#backward fill
d1.fillna(method='bfill')

Temperature  Humidity
0          1.0      22.0
1          3.0       2.0
2          3.0       2.0
3          2.0      20.0
4          3.0      20.0

Hey a fun fact here, as sweet as a cookie:

- ffill/pad/bfill are good imputation method if our data is of time series. This would keep the trend unaffected for our analysis.

Impute nan using interpolation method

In [ ]:
#interpolate
d1.interpolate()

Temperature  Humidity
0          1.0      22.0
1          2.0      12.0
2          3.0       2.0
3          2.0      11.0
4          3.0      20.0

You lucky champ! you got to know another amazing fact:
 - Interpolation method by default is linear in nature. It is an imputation technique that assumes a linear relationship between data points and utilises non-missing values from adjacent data points to compute a value for a missing data point.


You can explore other techniques involved in interplolation method, which might be usefull for your project.

Perform KNN imputation

Reference: https://link.medium.com/mV3023if8gb

In [ ]:
# Hint: Import KNNImputer and impute it on d1. Also note: Use n_neighbors=2
from sklearn.impute import KNNImputer
imputer = KNNImputer(n_neighbors=2)
d1_knn = pd.DataFrame(imputer.fit_transform(d1),columns=d1.columns)
d1_knn

Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  12.000000
4         3.00  20.000000

Point to ponder: KNN is an algorithm that is useful for matching a point with its closest k neighbors in a multi-dimensional space.

Do you think scaling is required to implement this method?. Yes you are right the answer is YES.
Can you comment below why normalized data is required, so that we understand your logic on this.

It requires us to normalize our data. Otherwise, the different scales of our data will lead the KNN Imputer to generate biased replacements for the missing values.

# Dropping Irrelevant Columns

Reference:-https://www.youtube.com/watch?v=cRurWEfmxC0

<p style='text-align: right;'> 5 points</p>


Create a dataframe df = pd.DataFrame(np.random.randint(0,100,size=(100, 5)), columns=list('ABCDE'))


In [ ]:
np.random.seed(10)
#define df
df = pd.DataFrame(np.random.randint(0,100,size=(100, 5)), columns=list('ABCDE')) 


print df

In [ ]:
df

A   B   C   D   E
0    9  15  64  28  89
1   93  29   8  73   0
2   40  36  16  11  54
3   88  62  33  72  78
4   49  51  54  77  69
..  ..  ..  ..  ..  ..
95   3  50  59  34  21
96  16  18  61  54  60
97  21  87  83  71  16
98  67  38  27  96  87
99  98  89  16  82  19

[100 rows x 5 columns]

Note: Since all the rows are having random numbers, your dataframe observations might be different than the output given above

Suppose for our analysis our project do not require column E. So you need to remove this column. update this new change using inplace parameter

In [ ]:
#drop E
df.drop('E',axis=1,inplace=True)

Check if column **E** is removed by printing head of df

In [ ]:
#df head
df.head()

A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77

# Ensure requirements as per domain

Reference:-https://www.youtube.com/watch?v=eM7p3MVLOZ8

<p style='text-align: right;'> 10 points</p>



Shallow copy the dataframe df in variable df2 and print df2 head






In [ ]:
df2=df.copy(deep=False)
df2
#df2 head


A   B   C   D
0    9  15  64  28
1   93  29   8  73
2   40  36  16  11
3   88  62  33  72
4   49  51  54  77
..  ..  ..  ..  ..
95   3  50  59  34
96  16  18  61  54
97  21  87  83  71
98  67  38  27  96
99  98  89  16  82

[100 rows x 4 columns]

Suppose your domain expert says to filter column B with even numbers to do correct analysis. Implement the same below and update the change in varaible df2.

In [ ]:
df2= df2[df2['B'] % 2 == 0]

print updated head of *df2*

In [ ]:
# df2 head
df2

A   B   C   D
2   40  36  16  11
3   88  62  33  72
6   30  30  89  12
9   11  28  74  88
10  15  18  80  71
14  88  50  54  34
15  77  88  15   6
18  97  50  45  40
23  81  80  41  90
26  96  66  67  62
30  88  96  73  40
32  28  30  89  25
34  33   4  87  94
37  68  30  70  74
38   9   2  65  13
40  62  42  34  40
41  32  94  86  58
42  45  18  50  44
45   6  44   9  50
46  44  68  14   4
49  39  58  81  39
51  69  48  60  58
54   5  70  61  16
55   4  22  19  85
58  10  36   0  30
59  85  32  81  13
60  31  32  96  38
62   0  96  25  63
63   2  30  17  59
64  63  86  29  71
65  19   0   7  93
67  58  76  10  79
70   7  88  96  38
71  27  64  49  90
72  27  52  90  57
76  99  46  46  45
77  84  20  49  33
79  77  66  37  31
85  82  56  41  90
86  74   8  92  32
87  87  68  86  14
90  50  50  85  14
92   1  28  58  10
93   1  82  34  11
94  74  36   6  63
95   3  50  59  34
96  16  18  61  54
98  67  38  27  96

# Creating a sensible index values

Reference:-https://www.youtube.com/watch?v=FQ4IB5mZOmQ


Oops. The index in this dataframe doesn't make sense. please correct index in a sequential manner starting from 1. Save the updates in df2

In [ ]:
#set proper index
new_index = [i+1 for i in range(len(df2))]
df2.set_index([new_index],inplace = True)


print df2 head again

In [ ]:
# df2 head
df2.head()

A   B   C   D
1  40  36  16  11
2  88  62  33  72
3  30  30  89  12
4  11  28  74  88
5  15  18  80  71

# Renaming column names to meaningful names.

Reference:- https://www.youtube.com/watch?v=YnggjqAfZCU
<p style='text-align: right;'> 2 points</p>


Now df2 columns represents marks of the adventurous 'Anand', the brave 'Barkha', the compassionate 'Chandu' and the dashing 'Daniel'. Rename the columns with their name inplace of their first letter of their name.

In [ ]:
#column renaming
df2.rename({'A':'Anand','B':'Barkha', 'C': 'Chandu', 'D':'Daniel'},axis=1,inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


print df2 tail with updated column names

In [ ]:
# df2 head
df2.tail()

Anand  Barkha  Chandu  Daniel
44      1      82      34      11
45     74      36       6      63
46      3      50      59      34
47     16      18      61      54
48     67      38      27      96

Yeah! now the data looks pretty meaningful to study

# Treating Duplicate Data

Reference: https://www.youtube.com/watch?v=bFVMR1qfzXo

<p style='text-align: right;'> 20 points</p>


Make another dataframe df3 by deep copying df2.

In [ ]:
df3=df2.copy(deep=True)

Make another column in df3 with name 'dummy' having 0 as values
throughout the rows.

In [ ]:
#assign dummy column full of zero value
df3['dummy'] = 0

In [ ]:
# print head of df, df2 and df3
print(df.head())
print(df2.head())
print(df3.head())


    A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77
   Anand  Barkha  Chandu  Daniel
1     40      36      16      11
2     88      62      33      72
3     30      30      89      12
4     11      28      74      88
5     15      18      80      71
   Anand  Barkha  Chandu  Daniel  dummy
1     40      36      16      11      0
2     88      62      33      72      0
3     30      30      89      12      0
4     11      28      74      88      0
5     15      18      80      71      0


Hey buddy! Don't you think, there is some difference between copy operation used for creating df2 and df3.

If you think Yes, Then please comment below the difference

In [ ]:
# comment

print tail of df3

In [ ]:
# df3 tail
df3.tail()

Anand  Barkha  Chandu  Daniel  dummy
44      1      82      34      11      0
45     74      36       6      63      0
46      3      50      59      34      0
47     16      18      61      54      0
48     67      38      27      96      0

make an array name 'ListB' with values of column 'Barkha'

In [ ]:
listB = np.array(df3.loc[:,'Barkha'])

Print ListB

In [ ]:
#print ListB
listB

array([36, 62, 30, 28, 18, 50, 88, 50, 80, 66, 96, 30,  4, 30,  2, 42, 94,
       18, 44, 68, 58, 48, 70, 22, 36, 32, 32, 96, 30, 86,  0, 76, 88, 64,
       52, 46, 20, 66, 56,  8, 68, 50, 28, 82, 36, 50, 18, 38])

Assign this array values as another column in df3 with name 'Anonymous'

In [ ]:
#create Anonymous column
df3['Anonymous'] = listB

Create a dataframe 'ListA' with values of row index 3, 10 and 40

In [ ]:
listA = df3.iloc[[2,9,39],:]

print ListA

In [ ]:
# print ListA
listA

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
3      30      30      89      12      0         30
10     96      66      67      62      0         66
40     74       8      92      32      0          8

Concat ListA to df3 ignoring the index values of ListA so that we can maintain the sequential index value thoughout the dataframe.

In [ ]:
df3 = pd.concat([df3,listA],ignore_index=True)

print head of df3

In [ ]:
# df3 head
df3.head()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
0     40      36      16      11      0         36
1     88      62      33      72      0         62
2     30      30      89      12      0         30
3     11      28      74      88      0         28
4     15      18      80      71      0         18

Check if there is any duplicate rows present in the dataframe df3

In [ ]:
#check duplicate
df3[df3.duplicated() != False]

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
48     30      30      89      12      0         30
49     96      66      67      62      0         66
50     74       8      92      32      0          8

By above output it seems we do have duplicated rows in our dataset

Drop duplicated rows using pandas function keeping first values of such duplicated observations

In [ ]:
#drop duplicate
df3.drop_duplicates(keep='first',inplace=True)

Check again if we have any duplicate row values present 

In [ ]:
#check duplicate
df3[df3.duplicated() != False]

Empty DataFrame
Columns: [Anand, Barkha, Chandu, Daniel, dummy, Anonymous]
Index: []

Yipeee!! Did you notice the dataframe is free from any duplicate rows now.

Drop any duplicate columns present in the dataframe df

In [ ]:
df3 = df.loc[:,~df.columns.duplicated()]

#print df3
df3

A   B   C   D
0    9  15  64  28
1   93  29   8  73
2   40  36  16  11
3   88  62  33  72
4   49  51  54  77
..  ..  ..  ..  ..
95   3  50  59  34
96  16  18  61  54
97  21  87  83  71
98  67  38  27  96
99  98  89  16  82

[100 rows x 4 columns]

Did you notice which Column is dropped?
I am sure you noticed it.
Name that column below

In [ ]:
#Column:___________ --> None

# Treating constant column values

<p style='text-align: right;'> 2 points</p>


Check unique values in each columns

In [ ]:
# df3 unique value
df3.nunique()

Anand     42
Barkha    32
Chandu    40
Daniel    36
dummy      1
dtype: int64

By above output which column has only 1 value as unique throught the rows? 
Yeah! you are right, its dummy column.
So lets drop it

Drop dummy column as it has constant values which will not give us any information and save the changes to df3 using inplace parameter

In [ ]:
# drop dummy
df3.drop(columns='dummy',axis=1,inplace = True)

print final obtained dataframe df3

In [ ]:
# print df3
df3

Anand  Barkha  Chandu  Daniel
1      40      36      16      11
2      88      62      33      72
3      30      30      89      12
4      11      28      74      88
5      15      18      80      71
6      88      50      54      34
7      77      88      15       6
8      97      50      45      40
9      81      80      41      90
10     96      66      67      62
11     88      96      73      40
12     28      30      89      25
13     33       4      87      94
14     68      30      70      74
15      9       2      65      13
16     62      42      34      40
17     32      94      86      58
18     45      18      50      44
19      6      44       9      50
20     44      68      14       4
21     39      58      81      39
22     69      48      60      58
23      5      70      61      16
24      4      22      19      85
25     10      36       0      30
26     85      32      81      13
27     31      32      96      38
28      0      96      25      63
29      2      30      17      59
30     63      86      29      71
31     19       0       7      93
32     58      76      10      79
33      7      88      96      38
34     27      64      49      90
35     27      52      90      57
36     99      46      46      45
37     84      20      49      33
38     77      66      37      31
39     82      56      41      90
40     74       8      92      32
41     87      68      86      14
42     50      50      85      14
43      1      28      58      10
44      1      82      34      11
45     74      36       6      63
46      3      50      59      34
47     16      18      61      54
48     67      38      27      96

# Iterating dataframes
<p style='text-align: right;'> 25 points </p>


Reference: https://www.youtube.com/watch?v=0nI3HTLPpZI
Let's look at three main ways to iterate over DataFrames.

1. iteritems()
2. iterrows()
3. itertuples()

We will also see time taken by these methods to print our dataframe. 

**1. Iterating DataFrames with iteritems()**

Lets iterate over rows of df3 uisng iteritems.



In [ ]:
import time
start = time.time()

#Use iteritems to iterate
for k,v in df3.iteritems():
  print(k, v)
print('Time taken(sec): ',(time.time()-start)*1000)

Anand 1     40
2     88
3     30
4     11
5     15
6     88
7     77
8     97
9     81
10    96
11    88
12    28
13    33
14    68
15     9
16    62
17    32
18    45
19     6
20    44
21    39
22    69
23     5
24     4
25    10
26    85
27    31
28     0
29     2
30    63
31    19
32    58
33     7
34    27
35    27
36    99
37    84
38    77
39    82
40    74
41    87
42    50
43     1
44     1
45    74
46     3
47    16
48    67
Name: Anand, dtype: int64
Barkha 1     36
2     62
3     30
4     28
5     18
6     50
7     88
8     50
9     80
10    66
11    96
12    30
13     4
14    30
15     2
16    42
17    94
18    18
19    44
20    68
21    58
22    48
23    70
24    22
25    36
26    32
27    32
28    96
29    30
30    86
31     0
32    76
33    88
34    64
35    52
36    46
37    20
38    66
39    56
40     8
41    68
42    50
43    28
44    82
45    36
46    50
47    18
48    38
Name: Barkha, dtype: int64
Chandu 1     16
2     33
3     89
4     74
5     80
6     54
7     15


Did you notice buddy how iteritems are iterating over df3.

Along with ways each iterating function works, also keep tallying the time taken for all other lopps too!. This will be fun, lets check iterrows()

**2. Iterating DataFrames with iterrows()**

In [ ]:
import time
start = time.time()
#Use iterrows to iterate
for k,v in df3.iterrows():
  print(k, v)
print('Time taken(sec): ',(time.time()-start)*1000)

1 Anand     40
Barkha    36
Chandu    16
Daniel    11
Name: 1, dtype: int64
2 Anand     88
Barkha    62
Chandu    33
Daniel    72
Name: 2, dtype: int64
3 Anand     30
Barkha    30
Chandu    89
Daniel    12
Name: 3, dtype: int64
4 Anand     11
Barkha    28
Chandu    74
Daniel    88
Name: 4, dtype: int64
5 Anand     15
Barkha    18
Chandu    80
Daniel    71
Name: 5, dtype: int64
6 Anand     88
Barkha    50
Chandu    54
Daniel    34
Name: 6, dtype: int64
7 Anand     77
Barkha    88
Chandu    15
Daniel     6
Name: 7, dtype: int64
8 Anand     97
Barkha    50
Chandu    45
Daniel    40
Name: 8, dtype: int64
9 Anand     81
Barkha    80
Chandu    41
Daniel    90
Name: 9, dtype: int64
10 Anand     96
Barkha    66
Chandu    67
Daniel    62
Name: 10, dtype: int64
11 Anand     88
Barkha    96
Chandu    73
Daniel    40
Name: 11, dtype: int64
12 Anand     28
Barkha    30
Chandu    89
Daniel    25
Name: 12, dtype: int64
13 Anand     33
Barkha     4
Chandu    87
Daniel    94
Name: 13, dtype: int64
14 A

**3. Iterating DataFrames with itertuples()**

In [ ]:
#iterate df3 using itertuples
import time
start = time.time()

#Use itertuples to iterate
for row in df3.itertuples():
  print(row)
print('Time taken(sec): ',(time.time()-start)*1000)

Pandas(Index=1, Anand=40, Barkha=36, Chandu=16, Daniel=11)
Pandas(Index=2, Anand=88, Barkha=62, Chandu=33, Daniel=72)
Pandas(Index=3, Anand=30, Barkha=30, Chandu=89, Daniel=12)
Pandas(Index=4, Anand=11, Barkha=28, Chandu=74, Daniel=88)
Pandas(Index=5, Anand=15, Barkha=18, Chandu=80, Daniel=71)
Pandas(Index=6, Anand=88, Barkha=50, Chandu=54, Daniel=34)
Pandas(Index=7, Anand=77, Barkha=88, Chandu=15, Daniel=6)
Pandas(Index=8, Anand=97, Barkha=50, Chandu=45, Daniel=40)
Pandas(Index=9, Anand=81, Barkha=80, Chandu=41, Daniel=90)
Pandas(Index=10, Anand=96, Barkha=66, Chandu=67, Daniel=62)
Pandas(Index=11, Anand=88, Barkha=96, Chandu=73, Daniel=40)
Pandas(Index=12, Anand=28, Barkha=30, Chandu=89, Daniel=25)
Pandas(Index=13, Anand=33, Barkha=4, Chandu=87, Daniel=94)
Pandas(Index=14, Anand=68, Barkha=30, Chandu=70, Daniel=74)
Pandas(Index=15, Anand=9, Barkha=2, Chandu=65, Daniel=13)
Pandas(Index=16, Anand=62, Barkha=42, Chandu=34, Daniel=40)
Pandas(Index=17, Anand=32, Barkha=94, Chandu=86, Dani

Hey buddy! so as you have seen every method works differently 
    
    iteritems(): Helps to iterate over each element of the set, column-wise. 
    iterrows(): Each element of the set, row-wise. 
    itertuple(): Each row and form a tuple out of them.

But if you ask for speed. The most best perfromance is given by itertuples compared to other two iterating methods.
So if anytime you need to save your computation time on iteration of dataframes you can go for itertuples. Was'nt it fun?:)

# Regular Expression 
<p style='text-align: right;'> 15 points </p>


Reference: https://www.youtube.com/watch?v=zTTQ8FE60j8

Reference: https://www.youtube.com/watch?v=h6E1PiTXnVI

Reference doc: https://www.guru99.com/python-regular-expressions-complete-tutorial.html

Python has a module named re to work with RegEx


### **!Are you ready to try regex on dataframes?**

*So here we go.!*

We are gonna try out following awesome re module functions

1. findall 
2. search
3. sub
4. split

If you want you can also refer the below regular expression syntax.

![image.png](attachment:image.png)


Hey future data scientists! we will now use regex on dataframes for data cleaning.

Who doesn't know Trump?. Lets dowload this interesting dataset of Trump insult tweets :https://www.kaggle.com/ayushggarg/all-trumps-twitter-insults-20152021/download

On this dataset we will learn how to use regex for data cleaning. By the way it will be also very usefull for feature engineering too!.



In [67]:
#load dataset

tweet_data= pd.read_csv(r'https://raw.githubusercontent.com/Sagar-Gilda/Python/main/trump_insult_tweets_2014_to_2021.csv')
tweet_data.tail()

Unnamed: 0  ...                                              tweet
10355       10356  ...  If Vice President @Mike_Pence comes through fo...
10356       10357  ...  States want to correct their votes, which they...
10357       10358  ...  They just happened to find 50,000 ballots late...
10358       10359  ...  The States want to redo their votes. They foun...
10359       10360  ...  Sleepy Eyes Chuck Todd is so happy with the fa...

[5 rows x 5 columns]

In [68]:
#import re module
import re

#print head of tweet_data
tweet_data.head()

Unnamed: 0  ...                                              tweet
0           1  ...  Can you believe this fool, Dr. Thomas Frieden ...
1           2  ...  Can you believe this fool, Dr. Thomas Frieden ...
2           3  ...  Big time in U.S. today - MAKE AMERICA GREAT AG...
3           4  ...  Politician @SenatorCardin didn't like that I s...
4           5  ...  For the nonbeliever, here is a photo of @Neily...

[5 rows x 5 columns]

Lets do some analysis using regex on this dataset

Before we go ahead, do you remember apply function? because you will have to require apply function to impliment regex methods.

You can refer this video so that you revisit how apply function works: https://www.youtube.com/watch?v=7HN-4Df8ZpA

**1. findall()**

Make another column 'year' with year in each row using regex on date column.


In [63]:
# create a function which takes date as parameter and applies regex on it
def pull_year(x):
  pattern = '\d+'
  year = re.findall(pattern,x)
  return year[0]

#use apply function on tweet_data to use above function in order to make year column
tweet_data['year'] = tweet_data['date'].apply(lambda x: pull_year(x))

#print tweet_data head
tweet_data.head()

date  ...  year
7621  2020-01-01  ...  2020
7622  2020-01-02  ...  2020
7623  2020-01-04  ...  2020
7624  2020-01-06  ...  2020
7625  2020-01-06  ...  2020

[5 rows x 5 columns]

lets filter year from 2020-2021 which was the election time in USA.

**2. search**

We will use regex search for this 

In [60]:
# create a function which takes year as parameter and applies regex on it
def search_year(x):
  year = re.search('(2020|2021)',x)
  return year

#use apply function on tweet_data to use above function in order to search tweets of 2020-2021
tweet_data['year'] = tweet_data['date'].apply(lambda x: search_year(x))

#print tweet_data head
tweet_data.head()

date  ...                                          year
7621  2020-01-01  ...  <re.Match object; span=(0, 4), match='2020'>
7622  2020-01-02  ...  <re.Match object; span=(0, 4), match='2020'>
7623  2020-01-04  ...  <re.Match object; span=(0, 4), match='2020'>
7624  2020-01-06  ...  <re.Match object; span=(0, 4), match='2020'>
7625  2020-01-06  ...  <re.Match object; span=(0, 4), match='2020'>

[5 rows x 5 columns]

You can also do the same thing using regex match function to do this which is vailable in pandas

Reference: https://www.geeksforgeeks.org/python-pandas-series-str-match/

In [80]:
# apply pandas str.match() function
tweet_data = tweet_data[tweet_data['date'].str.match(pat='(2020)|(2021)')]
tweet_data


date  ...         Name
7621   2020-01-01  ...  impeachment
7622   2020-01-02  ...  impeachment
7623   2020-01-04  ...  impeachment
7624   2020-01-06  ...          los
7625   2020-01-06  ...  impeachment
...           ...  ...          ...
10355  2021-01-06  ...         2020
10356  2021-01-06  ...         2020
10357  2021-01-06  ...         2020
10358  2021-01-06  ...         2020
10359  2021-01-06  ...        chuck

[2739 rows x 5 columns]

cool right!

You got some null values after applying above function. Lets drop them using dropna function. Also drop 'Unnamed: 0' column as it does not give any information.


In [70]:
#drop na and Unnamed: 0 column
tweet_data.dropna()
tweet_data.drop(columns=['Unnamed: 0'],inplace=True)
tweet_data

date  ...                                              tweet
0      2014-10-09  ...  Can you believe this fool, Dr. Thomas Frieden ...
1      2014-10-09  ...  Can you believe this fool, Dr. Thomas Frieden ...
2      2015-06-16  ...  Big time in U.S. today - MAKE AMERICA GREAT AG...
3      2015-06-24  ...  Politician SenatorCardin didn't like that I sa...
4      2015-06-24  ...  For the nonbeliever, here is a photo of Neilyo...
...           ...  ...                                                ...
10355  2021-01-06  ...  If Vice President Mike_Pence comes through for...
10356  2021-01-06  ...  States want to correct their votes, which they...
10357  2021-01-06  ...  They just happened to find 50,000 ballots late...
10358  2021-01-06  ...  The States want to redo their votes. They foun...
10359  2021-01-06  ...  Sleepy Eyes Chuck Todd is so happy with the fa...

[10360 rows x 4 columns]

**3. sub() Function**

Now you have filtered the dataset with 2739 rows. Let's remove all @ from tweet column suing re sub() function.

In [81]:
# create a function which takes tweet as parameter and applies regex on it
def remove_at(x):
  return re.sub('@','',x)

#use apply function on tweet_data to use above function in order to remove @ from tweets
tweet_data['tweet'] = tweet_data['tweet'].apply(lambda x: remove_at(x))
#tweet_data head
tweet_data

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


date  ...         Name
7621   2020-01-01  ...  impeachment
7622   2020-01-02  ...  impeachment
7623   2020-01-04  ...  impeachment
7624   2020-01-06  ...          los
7625   2020-01-06  ...  impeachment
...           ...  ...          ...
10355  2021-01-06  ...         2020
10356  2021-01-06  ...         2020
10357  2021-01-06  ...         2020
10358  2021-01-06  ...         2020
10359  2021-01-06  ...        chuck

[2739 rows x 5 columns]

You can also use the sub function just in one line using list comprehension. Can you try doing it below?

In [69]:
# sub() suing list comprehension
tweet_data['tweet'] = [re.sub('@','',tweet_data['tweet'][i]) for i in range(len(tweet_data))]
tweet_data

Unnamed: 0  ...                                              tweet
0               1  ...  Can you believe this fool, Dr. Thomas Frieden ...
1               2  ...  Can you believe this fool, Dr. Thomas Frieden ...
2               3  ...  Big time in U.S. today - MAKE AMERICA GREAT AG...
3               4  ...  Politician SenatorCardin didn't like that I sa...
4               5  ...  For the nonbeliever, here is a photo of Neilyo...
...           ...  ...                                                ...
10355       10356  ...  If Vice President Mike_Pence comes through for...
10356       10357  ...  States want to correct their votes, which they...
10357       10358  ...  They just happened to find 50,000 ballots late...
10358       10359  ...  The States want to redo their votes. They foun...
10359       10360  ...  Sleepy Eyes Chuck Todd is so happy with the fa...

[10360 rows x 5 columns]

**4. split() Function**

Lets now split the column target by making another column Name which has name before hyphen("-")


In [82]:
# create a function which takes target as parameter and applies regex on it
def split(x):
  Name = re.split("-",str(x))
  return Name[0]

#use apply function on tweet_data to use above function in order to create Name column
tweet_data['Name'] = tweet_data['target'].apply(lambda x: split(x))
#tweet_data head
tweet_data

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


date  ...         Name
7621   2020-01-01  ...  impeachment
7622   2020-01-02  ...  impeachment
7623   2020-01-04  ...  impeachment
7624   2020-01-06  ...          los
7625   2020-01-06  ...  impeachment
...           ...  ...          ...
10355  2021-01-06  ...         2020
10356  2021-01-06  ...         2020
10357  2021-01-06  ...         2020
10358  2021-01-06  ...         2020
10359  2021-01-06  ...        chuck

[2739 rows x 5 columns]

Now can you filter out Name which are specially targetted for trump? Lets do it below and check how many such tweets are there.

In [83]:
#filter Name which is euqals to trump
tweet_data[tweet_data['Name'] == 'trump']

date  ...   Name
7859   2020-02-09  ...  trump
7860   2020-02-09  ...  trump
7861   2020-02-09  ...  trump
7890   2020-02-12  ...  trump
7891   2020-02-12  ...  trump
...           ...  ...    ...
9658   2020-10-07  ...  trump
9659   2020-10-07  ...  trump
10091  2020-11-22  ...  trump
10188  2020-12-09  ...  trump
10258  2020-12-18  ...  trump

[65 rows x 5 columns]

So here you got total of 65 records which were  tweeted on Donald Trump in the span of 2020-2021. 

Well done buddy! You have learned how to apply regex on dataframes. Regex are mostly used for datasets which are having textual information.

Good job!  Now our interesting trump insult tweet data is somewhat cleaned. 

---------------------------------

# C'mon cheers:) you have completed the 5th milestone challenge too. 

--------------------------------

# FeedBack
We hope you’ve enjoyed this course so far. We’re committed to help you use "AI for All" course to its full potential, so that you have a great learning experience. And that’s why we need your help in form of a feedback here.

Please fill this feedback form https://docs.google.com/forms/d/e/1FAIpQLSfjBmH0yJSSA34IhSVx4h2eDMgOAeG4Dk-yHid__NMTk3Hq5g/viewform